src/pymor/algorithms/bfgs.py

```diff
@@ -0,0 +1,307 @@
+# This file is part of the pyMOR project (https://www.pymor.org).
+# Copyright pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
+
+import warnings
+
+import numpy as np
+
+from pymor.algorithms.line_search import armijo
+from pymor.core.defaults import defaults
+from pymor.core.exceptions import BFGSError
+from pymor.core.logger import getLogger
+from pymor.parameters.base import Mu
+
+
+def get_active_and_inactive_sets(parameter_space, mu, epsilon=1e-8):
+    """Compute the active and inactive parameter index sets for constrained optimization.
+
+    Parameters
+    ----------
+    parameter_space
+        The |ParameterSpace| used for checking parameter ranges.
+    mu
+        The parameter to compute the active and inactive sets for.
+    epsilon
+        Tolerance threshold for checking boundary conditions.
+
+    Returns
+    -------
+    active
+        The set of active parameter indices.
+    inactive
+        The set of inactive parameter indices.
+    """
+    mu = mu if isinstance(mu, Mu) else parameter_space.parameters.parse(mu)
+    active = np.array([])
+
+    for key, dim in parameter_space.parameters.items():
+        indices = np.zeros(dim)
+        low_range = parameter_space.ranges[key][0] * np.ones(dim)
+        high_range = parameter_space.ranges[key][1] * np.ones(dim)
+        eps = epsilon * np.ones(dim)
+        indices[np.where(np.logical_or(mu[key] - low_range <= eps, high_range - mu[key] <= eps))] = 1
+        active = np.concatenate((active, indices))
+
+    inactive = np.ones(active.shape) - active
+
+    return active, inactive
+
+
+def project_hessian(hessian, direction, active, inactive):
+    """Project the Hessian matrix using an active and an inactive set.
```

Comment:
The function name and also the description do not seem to fit to what the function does. It does not return a "projected Hessian", but the projection of the Hessian applied to the direction.

Reply:
I'll take a hint from Tim's repo and name it `compute_hessian_action_matrix`.